### ISSUE TITLE
Rotate Images

### ISSUE THREAD

**Issue by tiagoboldt** [fancyapps/fancyBox#1100]
I would very much appreciate if fancyBox was able to rotate the image being overlayed. I can image it as being an option such as:
```
<script>
    $(el).fancybox({
      'enableRotation': true,
    })
</script>
```
The option would place two buttons, to rotate the image +90º and -90º. It could also be made available in the buttons helper. 

Your thoughts?

**Comment by fancyapps** [comment]
Maybe in the future :)

**Comment by hiidemo** [comment]
Add this code to js file (fancybox 3)
Remmeber set "**instance.opts.rotate**" is true in variable **defaults** 

`// ==========================================================================
//
// Rotate image
// Adds rotate functionality
//
// ==========================================================================
;(function (document, $) {
	'use strict';

	var RotateImage = function( instance ) {

		this.instance = instance;

		this.init();

	};
	
	$.extend( RotateImage.prototype, {

		$button		: null,
		angle		: 0,

		init : function() {
			var self = this;

			self.$button = $('<button data-rotate-left class="fancybox-button fancybox-button--rotate" title="Xoay ảnh sang trái"><i class="fa fa-undo" aria-hidden="true"></i></button>')
				.appendTo( this.instance.$refs.buttons )
				.on('touchend click', function(e) {
					self.rotate('left');
				});
				
			self.$button = $('<button data-rotate-right class="fancybox-button fancybox-button--rotate" title="Xoay ảnh sang phải"><i class="fa fa-repeat" aria-hidden="true"></i></button>')
				.appendTo( this.instance.$refs.buttons )
				.on('touchend click', function(e) {
					self.rotate('right');
				});
		},

		rotate : function (direct) {
			var self = this;

			var image = self.instance.current.$image[0]; 
			var sign = (direct == 'right') ? "" : "-";
			
			self.angle += 90;
			
			$(image).css('webkitTransform', 'rotate(' + sign + self.angle + 'deg)');
			$(image).css('mozTransform', 'rotate(' + sign + self.angle + 'deg)');
		},
	});

	$(document).on('onInit.fb', function(e, instance) {
		if ( !!instance.opts.rotate) {
			instance.Rotate = new RotateImage( instance );
		}

	});

}(document, window.jQuery));`

**Comment by Wndworks** [comment]
@hiidemo Didn't work for me. I'm getting this error:

`Class 'Symfony\Component\HttpKernel\Exception\NotFoundHttpException' not found`

**Comment by Wndworks** [comment]
@hiidemo Didn't work. If I add your code to the Fancybox file and set "`instance.opts.rotate: true,`" in defaults, the image doesn't open in Fancybox anymore.

**Comment by JamesAndersonJr** [comment]
I think part of the problem may lie in the fact that the code was not posted correctly. Notice how
`}(document, window.jQuery));`
is not even in the code window?

**Comment by hiidemo** [comment]
Screenshot

![image](https://user-images.githubusercontent.com/12048174/34429796-9aab615c-ec8f-11e7-88da-9059fe6646a0.png)

**Comment by Wndworks** [comment]
@JamesAndersonJr Yes, i did notice that and fixed it from the beginning. I don't think that it's just a copy paste fail.

@hiidemo That's exactly what i'm trying to achieve. Can you help me out here? Is it possible to show the complete final script?

**Comment by paulhonnoll** [comment]
For fancybox 3 I did this....
`$(document).on('onInit.fb', function (e, instance) {
            var click = 1;
            $('.fancybox-toolbar').prepend('<button id="rotate_button" class="fancybox-button" title="Rotate Image"><i class="fa fa-repeat"></i></button>')
                .on('click', '#rotate_button', function () {
                    var n = 90 * ++click;
                    $('.fancybox-image-wrap img').css('webkitTransform', 'rotate(-' + n + 'deg)');
                    $('.fancybox-image-wrap img').css('mozTransform', 'rotate(-' + n + 'deg)');
                });
        });`

**Comment by Wndworks** [comment]
@paulhonnoll That's a nice solution! Thank you!

**Comment by JamesAndersonJr** [comment]
@paulhonnoll Great Job! I have just 1 recommendation: change your preceding comment to block style, for copy-and-pasting. I.e.:

`/* adding custom item to fancybox menu to rotate image */`

**Comment by paulhonnoll** [comment]
Thanks @JamesAndersonJr , made the update!

**Comment by Eldenroot** [comment]
Really nice addition, would be useful :)

**Comment by paulhonnoll** [comment]
I tested it with Firefox, IE 10+ and Chrome and it was working.

**Comment by acwolff** [comment]
Wij don't rotate you your images in a photo editor before you upload the images to a web sertver?

**Comment by Wndworks** [comment]
@Eldenroot i still use it, works perfect.

@acwolff Because in some cases you have to deal with website users that are uploading the images

**Comment by acwolff** [comment]
@Wndworks in that case, the up-loaders should rotate the images in a photo editor before they upload!
I think it is not worthwhile to add code to FancyBox for this purpose!

**Comment by Eldenroot** [comment]
@Wndworks - unfortunately not working for me, can you post your working code on codepen? thank you!

**Comment by seltix5** [comment]
Unbelivebel this feature is not more requested :o

**Comment by seltix5** [comment]
hello all,
i updated @hiidemo code and its working with the current fancybox version (3.5.6) :

```
var RotateImage = function (instance) {
    this.instance = instance;

    this.init();
};

$.extend(RotateImage.prototype, {
    $button: null,
    angle: 0,

    init: function () {
        var self = this;

        self.$button = $('<button data-rotate-left class="fancybox-button fancybox-button--rotate" title="Rotate to left"><i class="fas fa-undo" aria-hidden="true"></i></button>')
            .prependTo(this.instance.$refs.toolbar)
            .on('touchend click', function (e) {
                self.rotate('left');
            });

        self.$button = $('<button data-rotate-right class="fancybox-button fancybox-button--rotate" title="Rotate to right"><i class="fas fa-redo" aria-hidden="true"></i></button>')
            .prependTo(this.instance.$refs.toolbar)
            .on('touchend click', function (e) {
                self.rotate('right');
            });
    },

    rotate: function (direction) {
        var self = this;

        var image = self.instance.current.$image[0];
        if (direction == 'right') {
            self.angle += 90;
        } else {
            self.angle -= 90;
        }


        $(image).css('transition', 'transform .3s ease-in-out');

        $(image).css('webkitTransform', 'rotate(' + self.angle + 'deg)');
        $(image).css('mozTransform', 'rotate(' + self.angle + 'deg)');
    }
});

$(document).on('onInit.fb', function (e, instance) {
    if (!!instance.opts.rotate) {
        instance.Rotate = new RotateImage(instance);
    }
});
```
The code can be added at the end of the "jquery.fancybox.js" file or just loaded after the file.

The icons are using last version of font awesome (v5).

Updates:
- the current version of fancybox uses "this.instance.$refs.toolbar" instead of "this.instance.$refs.buttons"
- changed the "appendTo" to "prependTo" to add the icons at left of the existent ones
- updated the rotate counter since las one was not working correctly
- added transition css

**Comment by Eldenroot** [comment]
Thank you!

**Comment by Eldenroot** [comment]
@seltix5 - is possible to replace FA5 icons by svg?

**Comment by seltix5** [comment]
@Eldenroot Hello, yes, I supose. You probably will need the FA API for that since the svg is generated from JS and replace the `<i>` icons with the var with the generated icon :

`var igloo = window.FontAwesome.icon({ prefix: 'fas', iconName: 'igloo' })`

https://fontawesome.com/how-to-use/with-the-api/setup/getting-started

let me know if it works!

**Comment by Eldenroot** [comment]
I mean use same way as FancyBox does for its icons, not FA

**Comment by seltix5** [comment]
Yes, you need to generate the svg code you want and put it there instead of the `<i>`

**Comment by Eldenroot** [comment]
do you have any generated example which fits for this purpose?

**Comment by seltix5** [comment]
I do not since I use FA icons. I will need to seach online for a vector icon or create one and export it as SVG or use online converter with a vector image

**Comment by Eldenroot** [comment]
that would be great!

**Comment by seltix5** [comment]
even more simple, you could just go to the FA icons page and copy the svg code from them =)

**Comment by Eldenroot** [comment]
hmm, working example please? it is only for PRO not?

**Comment by Eldenroot** [comment]
Moreover would be nice to add support to just customize the output - I mean just add "rotate" into buttons setting to display or hide the button :)

**Comment by acwolff** [comment]
Just Google: 
https://www.google.nl/search?tbm=isch&source=hp&ei=hOmLXPDnFIe3kwXxmKgo&q=rotate+icon+svg&oq=svg+rotate&gs_l=mobile-gws-wiz-img.1.1.0i19j0i8i30i19l3.4108.9608..15651...0.0..0.61.535.11......0....1.......0..0.vHV1hdYgUh4

**Comment by cemilf** [comment]
> do you have any generated example which fits for this purpose?

@Eldenroot You can use the following code for buttons containing SVG icons

```
self.$button_right = $('<button data-rotate-right class="fancybox-button fancybox-button--rotate" title="Rotate to right">' +
    '<svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24">' +
    '  <path d="M11.074,9.967a4.43,4.43,0,1,1-4.43-4.43V8.859l5.537-4.43L6.644,0V3.322a6.644,6.644,0,1,0,6.644,6.644Z" transform="translate(10.305 1) rotate(30)"/>' +
    '</svg>' +
    '</button>')
    .prependTo(this.instance.$refs.toolbar)
    .on('click', function (e) {
        self.rotate('right');
    });

self.$button_left = $('<button data-rotate-left class="fancybox-button fancybox-button--rotate" title="Rotate to left">' +
        '<svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24">' +
        '  <path d="M11.074,6.644a4.43,4.43,0,1,0-4.43,4.43V7.752l5.537,4.43-5.537,4.43V13.289a6.644,6.644,0,1,1,6.644-6.644Z" transform="translate(21.814 15.386) rotate(150)"/>' +
        '</svg>' +
        '</button>')
    .prependTo(this.instance.$refs.toolbar)
    .on('click', function (e) {
        self.rotate('left');
    });                
```

**Comment by liqinan158** [comment]
@cemilf , Thanks for your comments. Most of the time it's working. While it can not work when the image ratio is not 1, the movement of the rotated image will be restricted to the raw container whose size is not matching.

**Comment by liqinan158** [comment]
@fancyapps Dear devs, any news of this function?

**Comment by Eldenroot** [comment]
I think that FancyBox is not dev's priority right now... it is not good, because it is a really cool library.

**Comment by winyle-online** [comment]
Thanks @hiidemo and @seltix5 for code. That was exactly what I was looking for. I added also an initial rotation of specified degree (set in database) after open image preview as some of my scans have wrong rotation.

**Comment by fancyapps** [comment]
I can inform you that Fancybox v5 will provide not only functionality to rotate images, but also to flip vertically / horizontally (and to skew, if needed 😃). Sorry, I can not give any ETA. I have now finished working on the Panzoom component (which is basically the backbone of Carousel and Fancybox) and I am now working to implement the most requested functionality for the Carousel.

**Comment by Eldenroot** [comment]
> I can inform you that Fancybox v5 will provide not only functionality to rotate images, but also to flip vertically / horizontally (and to skew, if needed 😃). Sorry, I can not give any ETA. I have now finished working on the Panzoom component (which is basically the backbone of Carousel and Fancybox) and I am now working to implement the most requested functionality for the Carousel.

This sounds great! Thank you @fancyapps

**Comment by fancyapps** [comment]
Well, sorry, migrating the whole project to TS and implementing the many requests is a huge amount of work, but now v5 is nearing completion.
